budgetkey_data_pipelines/pipelines/budget/national/changes/processed/extract-change-groups.py

```diff
@@ -0,0 +1,237 @@
+    new_fields.append({
+        'name': 'transaction_id',
+        'type': 'string'
+    })
+    resource['schema']['fields'] = new_fields
+    return datapackage
+
+
+def transfer_code(change):
+    return '%d/%02d-%03d' % (
+        change['year'],
+        change['leading_item'],
+        change['req_code']
+    )
+
+
+def powerset(iterable):
+    s = list(iterable)
+    return itertools.chain.from_iterable(
+        itertools.combinations(s, r)
+        for r in range(2, min(len(s) + 1, 6))
+    )
+
+
+def subsets(s):
+    return map(set, powerset(s))
+
+
+def get_changes(rows):
+    now = datetime.now().strftime('%d/%m/%Y')
+    for row in rows:
+        row['trcode'] = transfer_code(row)
+
+        if row['date/approval'] is None:
+            row['date_kind'] = 'pending/' + now
+        else:
+            row['date_kind'] = 'approved/' + (
+                row['date/approval'].strftime('%d/%m/%Y')
+            )
+
+        yield row
+
+
+def get_transactions(changes):
+    def get_date(x):
+        return x['date_kind']
+
+    def get_keys_for_prefixes(ch, prefixes):
+        result = set()
+
+        for c in ch:
+            key = (c['leading_item'], c['req_code'])
+            if key not in result:
+                for prefix in prefixes:
+                    if c['budget_code'].startswith(prefix):
+                        result.add(key)
+                        break
+
+        return result
+
+    def find_groups(changes):
+        selected_transfer_codes = set()
+        for date_kind, date_changes in itertools.groupby(changes, get_date):
+            date_reserve = [
+                c for c in date_changes
+                if c['budget_code'].startswith('0047') and c['leading_item'] != 47
+                if any(c[field] for field in value_fields)
+            ]
+            logging.debug('Reserve date: kind:%s num:%s' % (
+                date_kind, len(date_reserve)
+            ))
+            num_found = 0  # these two - only to show progress
+            i = 0          #
+            for comb_size in range(2, min(len(date_reserve) + 1, 7)):
+                done = False
+                while not done:
+                    not_selected = list(
+                        x for x in date_reserve
+                        if x['trcode'] not in selected_transfer_codes
+                    )
+                    logging.debug('len(not_selected)=%d' % len(not_selected))
+                    date_transactions = itertools.combinations(
+                        not_selected, comb_size)
+                    found = None
+                    done = True
+                    try:
+                        while True:
+                            i += 1
+                            if i % 100000 == 0:
+                                logging.debug('%s %s %s' % (
+                                    len(date_reserve), num_found, i
+                                ))
+                            transaction = date_transactions.send(found)
```

Comment:
Yes - in the `combinations` version that was removed, there was logic to allow the receiver to send the generator an indication if a match was found.
Then, the generator (i.e. the `combinations` function) removed all found elements from the pool, this decreasing the number of combinations.
For example, if `combinations` returned [1,2,5] and there was a match - then there's no point in testing [1,2,6] as well.